modal/lerobot_modal/conversion_service.py

```diff
@@ -0,0 +1,95 @@
+import modal
+
+
+conversion_image = (
+    modal.Image.debian_slim(python_version="3.10")
+    .apt_install(
+        "ffmpeg",
+        "libavutil-dev",
+        "libavcodec-dev",
+        "libavformat-dev",
+    )
+    .env(
+        {
+            "HF_HUB_ENABLE_HF_TRANSFER": "1",
+            "HF_HUB_DISABLE_TELEMETRY": "1",
+        }
+    )
+    .uv_pip_install(
+        "lerobot==0.4.0",  # This version includes v3.0 of LeRobotDataset
+        "loguru",
+        "huggingface_hub[cli]",
+    )
+)
+
+conversion_service = modal.App("conversion-app")
+
+
+@conversion_service.function(
+    image=conversion_image,
+    cpu=1,  # Single core is enough from experience
+    timeout=15 * 60,
+    secrets=[
+        modal.Secret.from_dict({"MODAL_LOGLEVEL": "DEBUG"}),
+        modal.Secret.from_name("huggingface"),
+    ],
+)
+async def convert_dataset_to_v3(
+    dataset_name: str,
+    huggingface_token: str | None = None,
+) -> tuple[str | None, str | None]:
+    """
+    Convert a v2.1 dataset to LeRobot v3.0 format and upload to Hugging Face Hub.
+    If a token is passed, we assume it has the necessary permissions to push to the dataset.
+    If not, we reupload a version of the dataset on the phospho cloud.
+    """
+    import os
+    from loguru import logger
+    from huggingface_hub import (
+        snapshot_download,
+        upload_folder,
+        create_repo,
+        create_tag,
+    )
+    from lerobot.datasets.v30.convert_dataset_v21_to_v30 import convert_dataset
+
+    try:
+        # Clear Hugging Face cache
+        os.system("huggingface-cli delete-cache -y")
+
+        # We do this because LeRobot later uses HfApi internally which reads from env variables
+        if huggingface_token is not None:
+            os.environ["HF_TOKEN"] = huggingface_token
+        else:
+            if dataset_name.startswith("phospho-app/"):
+                # Dataset is already on our account, no need to reupload
+                pass
+        # In this case, we need to reupload the dataset on our account to have write permissions
+        dataset_path_as_str = snapshot_download(
+            repo_id=dataset_name, repo_type="dataset", revision="v2.1"
+        )
+        new_repo = "phospho-app/" + dataset_name.split("/")[-1]
```

Comment:
I disagree with this logic. If the user is in private mode or if we have their token, we should push the data back to the same dataset, just on a new branch.